src/Microsoft.Health.Fhir.Liquid.Converter/Filters/GeneralFilters.cs

```diff
@@ -53,26 +53,37 @@ public static string Evaluate(string input, string property)
 
         public static string GenerateIdInput(Context context, string segment, string resourceType, bool isBaseIdRequired, string baseId = null)
         {
-            if (string.IsNullOrEmpty(segment))
+            if (string.IsNullOrWhiteSpace(segment))
             {
-                return segment;
+                return null;
             }
 
             if (string.IsNullOrEmpty(resourceType) || (isBaseIdRequired && string.IsNullOrEmpty(baseId)))
             {
                 throw new RenderException(FhirConverterErrorCode.InvalidIdGenerationInput, Resources.InvalidIdGenerationInput);
             }
 
-            var input = baseId != null ? $"{baseId}_{resourceType}_{segment}" : $"{resourceType}_{segment}";
+            // Normalize encoding characters
             var encodingCharacters = (Hl7v2EncodingCharacters)context["EncodingCharacters"];
-            return NormalizeIdInput(input, encodingCharacters);
+            var separators = $"{encodingCharacters.ComponentSeparator}{encodingCharacters.RepetitionSeparator}{encodingCharacters.EscapeCharacter}{encodingCharacters.SubcomponentSeparator}";
+            var formattedSegment = segment.Trim().Replace(separators, ",,,,");
+
+            formattedSegment = SegmentSeparatorsRegex.Replace(formattedSegment, ",");
+            formattedSegment = formattedSegment.Replace(encodingCharacters.FieldSeparator, ',')
+                .Replace(encodingCharacters.ComponentSeparator, ',')
+                .Replace(encodingCharacters.SubcomponentSeparator, ',')
+                .Replace(encodingCharacters.RepetitionSeparator, ',');
+
+            formattedSegment = SpecialCharProcessor.Escape(Hl7v2EscapeSequenceProcessor.Unescape(formattedSegment, encodingCharacters));
+
+            return baseId != null ? $"{baseId}_{resourceType}_{formattedSegment}" : $"{resourceType}_{formattedSegment}";
         }
 
         public static string GenerateUUID(string input)
         {
             if (string.IsNullOrEmpty(input))
```

Comment:
If the `input` is composed with whitespaces, we will still generate UUID for it, right?

Reply:
Whitespaces will be transformed to null in `GenerateIdInput`, which is called before this function. But we can still check whitespaces here in case users call this function directly. I'll fix it in the update. Thanks!

Reply:
Got it, thanks.